### DISCUSSION TITLE
Nuxt Plugin: navigation guard

### DISCUSSION THREAD

**Discussion by elHornair** [nuxt/nuxt#18031] [Questions] [answered]
Hello👋 

I'm doing this:

```js
export default defineNuxtPlugin((nuxtApp) => {
    console.log("Plugin initialised");

    nuxtApp.$router.beforeEach((to) => {
        console.log("beforeEach called");
    });
});
```

The plugin gets initialised on both server and client side.
On the server side, the navigation guard is called on initial page load.
On the client side then, the navigation guard is NOT called on initial page load. Why? Am I missing something fundamental?

**Comment by Diizzayy** [comment, marked as answer]
Okay so, firstly let's clear up a few things.

> On the server side, the navigation guard is called on initial page load.

The function is indeed called, however though you'll find that the currentRoute is not correctly propagated to reflect the page that you're actually viewing. ie: the path always resolves to '/' ( at this point in Nuxt3's beta atleast ) which makes it very useless for navigation guards.

> On the client side then, the navigation guard is NOT called on initial page load

Currently this behavior might be a bug as far as I know.

Now, as for how a simplified version of how I've been implementing Navigation Guards in Nuxt3, here goes:

```js
export default defineNuxtPlugin(({ ssrContext }) => {
  const router = useRouter();

  if (process.server && ssrContext) {
    const { res, url } = ssrContext;
    const { path } = router.resolve(url);

    if (path === "/private") {
      res.writeHead(302, { Location: "/" });
      res.end();
    }
  } else if (process.client) {
    router.beforeEach((to, _, next) => {
      if (to.path === "/private") return next("/");

      return next();
    });
  }
});
```

This can easily be modified to account for authenticated user roles, etc... anything your authentication logic allows for.

**Reply by elHornair** [reply]
You, my friend, are a hero 🚀 Thanks so much!

**Reply by vincenthalman** [reply]
I was struggling with the same issue.. What I did was to use beforeRouteEnter on the page that needed some auth ( as a quick fix till nuxt3 got some better navigation guard fix, this also works on page reload ).


    beforeRouteEnter(to, from, next) {
      const store = userStore()
      const unwrap_storeuser = unref(store.user)
      const u = JSON.parse(JSON.stringify(unwrap_storeuser))


      if (!u.isAdmin) { // if user is not admin, return
        next('/admin/login')
      } else {
        next()
      }
    }
